### PR TITLE
feat: added support for hf assistant mask and updated regex detection with fallback

### LINKED ISSUE #210: Update assistant pattern detection logic
The `_detect_assistant_pattern` function in `src/speculators/data_generation/preprocessing.py` dynamically detects the assistant message pattern from a tokenizer's chat template, but returns a regex that
  contains a hardcoded negative lookahead pattern `(?!<\|start\|)` on line 145. This hardcoded pattern assumes all chat templates use `<|start|>` as a role boundary marker, which is specific to certain tokenizers
   (e.g., GPT-OSS, Llama) but may not work for other chat template formats. The negative lookahead should be dynamically extracted from the detected chat template structure instead of being hardcoded,
  ensuring the function works correctly across all tokenizer chat templates. Currently, the function successfully detects the role marker prefix and suffix dynamically but fails to apply the same approach to
  the boundary prevention pattern, limiting its generalizability.

**TODO:**
  - Dynamically extract the role boundary marker (e.g., <|start|>) from the chat template instead of hardcoding it in the negative lookahead.
  - Consider leveraging HuggingFace's `apply_chat_template(..., return_assistant_tokens_mask=True)` parameter as an alternative approach for identifying assistant message spans, which would eliminate the need for regex pattern matching entirely and provide a more robust, template-agnostic solution. We have previously explored this solution but seems like it only supports limited models.
  - Add unit tests covering multiple chat template formats to ensure the pattern detection works universally.

### PR DESCRIPTION
**Pull request by SwekeR-463**
Fixes #210 

- Made changes to `src/speculators/datagen/preprocessing.py` for assistant mask and updated the regex detection while ensuring fallback logic exists when assistant mask is not present in tokenizer config.
- Added 2 unit tests to `tests/datagen/test_preprocessing.py` for handling the made changes for assistant mask.

Implemented:
- [x] Dynamically extract the role boundary marker (e.g., <|start|>) from the chat template instead of hardcoding it in the negative lookahead.
- [x] Consider leveraging HuggingFace's apply_chat_template(..., return_assistant_tokens_mask=True) parameter as an alternative approach for identifying assistant message spans, which would eliminate the need for regex pattern matching entirely and provide a more robust, template-agnostic solution.
- [x] Add unit tests covering multiple chat template formats to ensure the pattern detection works universally.

### CODE REVIEW

File: src/speculators/data_generation/preprocessing.py [hunk 1]
```diff
@@ -271,10 +316,14 @@ def build_eagle3_dataset(
...
+    # Detect and use provided assistant message pattern
+    supports_mask = _supports_assistant_mask(tokenizer)
+
+    if assistant_pattern is None and not supports_mask:
         assistant_pattern = _detect_assistant_pattern(tokenizer)
         log.info(f"Detected assistant pattern: {assistant_pattern[:80]}...")
+    elif supports_mask:
+        log.info("Using HF assistant token mask for loss masking")
     else:
         log.info(f"Using custom assistant pattern: {assistant_pattern[:80]}...")
```
Comment (fynnsu) [inline src/speculators/data_generation/preprocessing.py]: I think it would be better if we have the following flow (pseudocode):

```python
if assistant_pattern is not None:
    # Use assistant pattern
elif _supports_assistant_mask(tokenizer):
    # Use tokenizer w/ `return_assistant_tokens_mask=True`
else:
    assistant_pattern = _detect_assistant_pattern(tokenizer)
```

This way users can still specify an `assistant_pattern` to override the `return_assistant_tokens_mask` logic if they choose to.


The other change required for this is to make `_preprocess_batch` accept `str | None` for the assistant pattern and then we can assume we're using the tokenizer's automatic detection if the assistant pattern is `None`.

File: src/speculators/data_generation/preprocessing.py [hunk 2]
```diff
@@ -196,6 +224,9 @@ def _preprocess_batch(
         log.warning(f"No conversations key found. Keys: {list(examples.keys())}")
         return results
 
+    # Check for mask support once per batch
+    supports_mask = _supports_assistant_mask(tokenizer)
+
```
Comment (fynnsu) [inline src/speculators/data_generation/preprocessing.py]: Read other comment first.

We don't need this check if we can allow `assistant_pattern` to be `None` and treat that as an indicator of whether the tokenizer accepts `return_assistant_tokens_mask=True` or not. This is also good because we probably don't want to be doing this check for every single batch.

File: src/speculators/data_generation/preprocessing.py [hunk 3]
```diff
@@ -140,9 +153,24 @@ def _detect_assistant_pattern(tokenizer: PreTrainedTokenizer) -> str:
     # Extract suffix from the last assistant message
     suffix = formatted[second_end:]
 
-    # Use negative lookahead to prevent matching across role boundaries
-    # This prevents the pattern from spanning multiple messages
-    return re.escape(role_marker) + r"((?:(?!<\|start\|).)*?)" + re.escape(suffix)
+    # Extract dynamic boundary marker from role_marker
+    boundary_match = re.search(
+        r"((<\|?[a-zA-Z0-9_]+[\|>]?)|(\[[a-zA-Z0-9_]+\]))", role_marker
```
Comment (shanjiaz) [inline src/speculators/data_generation/preprocessing.py:197]: Can we make sure this regex works with patterns such as `<|im_start|>` , `[/INST]`.

**Review by shanjiaz — COMMENTED** [review]
This looks great! Thanks for adding the tests as well. It would be awesome if you could share some test results showing the regex working correctly across a few different model families (Llama, Qwen, Mistral, Gemma, etc.). This would really validate the template-agnostic design and might catch any edge cases! Thanks again for this awesome contribution.

**Comment by SwekeR-463** [comment]
@shanjiaz pasting test results for different model families
models used - qwen2, llama3, mistral, gemma, phi3 & gpt-oss-20b
also have added the test into `tests/datagen/test_regex_patterns.py`
```bash
test_regex_patterns.py::test_regex_detection_across_models[Qwen/Qwen2-0.5B-Instruct] 
Testing family: Qwen/Qwen2-0.5B-Instruct
  Detected pattern: <\|im_start\|>assistant\
((?:(?!<\|im_start\|).)*?)<\|im_end\|>\

  Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
PASSED
---------------------------------------------------------------------------------------------------------------------------------------------------------
test_regex_patterns.py::test_regex_detection_across_models[unsloth/llama-3-8b-Instruct] 
Testing family: unsloth/llama-3-8b-Instruct
  Detected pattern: <\|start_header_id\|>assistant<\|end_header_id\|>\
\
((?:(?!<\|start_header_id\|).)*?)<\|eot_id\|>
  Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
PASSED
-------------------------------------------------------------------------------------------------------------------------------------------------------
test_regex_patterns.py::test_regex_detection_across_models[mistralai/Mistral-7B-Instruct-v0.2] 
Testing family: mistralai/Mistral-7B-Instruct-v0.2
  Detected pattern: \ \[/INST\]\ ((?:(?!<\|start\|).)*?)</s>
  Decoded trainable regions: I am a helpful assistant. The capital of France is Paris.
PASSED
------------------------------------------------------------------------------------------------------------------------------------------------------------
test_regex_patterns.py::test_regex_detection_across_models[google/gemma-2b-it] 
Testing family: google/gemma-2b-it
  Detected pattern: <end_of_turn>\
<start_of_turn>model\
((?:(?!<end_of_turn>).)*?)<end_of_turn>\

  Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
PASSED
--------------------------------------------------------------------------------------------------------------------------------------------------------
test_regex_patterns.py::test_regex_detection_across_models[microsoft/Phi-3-mini-4k-instruct] 
Testing family: microsoft/Phi-3-mini-4k-instruct
  Detected pattern: <\|assistant\|>\
((?:(?!<\|assistant\|).)*?)<\|end\|>\
<\|
  Decoded trainable regions: I am a helpful assistant. The capital of France is Paris.
PASSED
--------------------------------------------------------------------------------------------------------------------------------------------------------
test_regex_patterns.py::test_regex_detection_across_models[openai/gpt-oss-20b] 
Testing family: openai/gpt-oss-20b
  Detected pattern: <\|start\|>assistant<\|channel\|>final<\|message\|>((?:(?!<\|start\|).)*?)<\|
  Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
PASSED
```

File: src/speculators/data_generation/preprocessing.py (hunk 2 above)
Comment (SwekeR-463) [reply src/speculators/data_generation/preprocessing.py]: have made the changes accordingly and have pushed the changes

**Comment by shanjiaz** [comment]
> @shanjiaz pasting test results for different model families models used - qwen2, llama3, mistral, gemma, phi3 & gpt-oss-20b also have added the test into `tests/datagen/test_regex_patterns.py`
> 
> ```shell
> test_regex_patterns.py::test_regex_detection_across_models[Qwen/Qwen2-0.5B-Instruct] 
> Testing family: Qwen/Qwen2-0.5B-Instruct
>   Detected pattern: <\|im_start\|>assistant\
> ((?:(?!<\|im_start\|).)*?)<\|im_end\|>\
> 
>   Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
> PASSED
> ---------------------------------------------------------------------------------------------------------------------------------------------------------
> test_regex_patterns.py::test_regex_detection_across_models[unsloth/llama-3-8b-Instruct] 
> Testing family: unsloth/llama-3-8b-Instruct
>   Detected pattern: <\|start_header_id\|>assistant<\|end_header_id\|>\
> \
> ((?:(?!<\|start_header_id\|).)*?)<\|eot_id\|>
>   Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
> PASSED
> -------------------------------------------------------------------------------------------------------------------------------------------------------
> test_regex_patterns.py::test_regex_detection_across_models[mistralai/Mistral-7B-Instruct-v0.2] 
> Testing family: mistralai/Mistral-7B-Instruct-v0.2
>   Detected pattern: \ \[/INST\]\ ((?:(?!<\|start\|).)*?)</s>
>   Decoded trainable regions: I am a helpful assistant. The capital of France is Paris.
> PASSED
> ------------------------------------------------------------------------------------------------------------------------------------------------------------
> test_regex_patterns.py::test_regex_detection_across_models[google/gemma-2b-it] 
> Testing family: google/gemma-2b-it
>   Detected pattern: <end_of_turn>\
> <start_of_turn>model\
> ((?:(?!<end_of_turn>).)*?)<end_of_turn>\
> 
>   Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
> PASSED
> --------------------------------------------------------------------------------------------------------------------------------------------------------
> test_regex_patterns.py::test_regex_detection_across_models[microsoft/Phi-3-mini-4k-instruct] 
> Testing family: microsoft/Phi-3-mini-4k-instruct
>   Detected pattern: <\|assistant\|>\
> ((?:(?!<\|assistant\|).)*?)<\|end\|>\
> <\|
>   Decoded trainable regions: I am a helpful assistant. The capital of France is Paris.
> PASSED
> --------------------------------------------------------------------------------------------------------------------------------------------------------
> test_regex_patterns.py::test_regex_detection_across_models[openai/gpt-oss-20b] 
> Testing family: openai/gpt-oss-20b
>   Detected pattern: <\|start\|>assistant<\|channel\|>final<\|message\|>((?:(?!<\|start\|).)*?)<\|
>   Decoded trainable regions: I am a helpful assistant.The capital of France is Paris.
> PASSED
> ```

This looks great! Thanks for adding the unit test!

**Review by shanjiaz — DISMISSED** [review]
Looks great!

**Review by fynnsu — DISMISSED** [review]
Thank you for making the changes!

**Comment by SwekeR-463** [comment]
just had one doubt why the build run fails 
when i raised pr previously also the build run failed too 
do i have to do some change from my end 
also changing gated gemma model to unsloth one

File: tests/datagen/test_regex_patterns.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+"""
+
+import sys
+from pathlib import Path
+from re import Pattern
+
+import pytest
+from transformers import AutoTokenizer
+
+sys.path.append(str(Path(__file__).resolve().parents[2] / "src"))
```
Comment (fynnsu) [inline tests/datagen/test_regex_patterns.py]: Could you remove this line, it's actually not needed since the package is already pip installed.

**Review by fynnsu — COMMENTED** [review]
> why the build run fails

Yes, sorry you can ignore that. There's an issue with the build script when prs are opened from forks. We will fix this but it won't be a blocker for this pr.

I did see one other small thing in the tests, but otherwise this should be good to go.

File: tests/datagen/test_regex_patterns.py (hunk 4 above)
Comment (SwekeR-463) [reply tests/datagen/test_regex_patterns.py]: done :+1:

**Review by fynnsu — APPROVED** [review]

**Review by shanjiaz — APPROVED** [review]
